### PR TITLE
Ghe URI support fo RtGithub

### PR DESCRIPTION
**Pull request by perfa**
This adds a few constructors to `RtGithub` in order to simplify creation when using jcabi with a github enterprise deployment. While the `Request` constructor does allow use with GHE, as a user you have to look up how the default `Request` object is created and copy that. Uses java.net's `URI` object to make the signature unique versus the user:password and token constructors.

### CODE REVIEW

**Comment by dmarkov** [comment]
@perfa Thanks, someone will review your pull request soon

**Comment by dmarkov** [comment]
@pinaf please review this PR

File: src/test/java/com/jcabi/github/RtGithubTest.java [hunk 1]
```diff
@@ -191,4 +193,18 @@ public void equalsToAnotherGithub() throws Exception {
             Matchers.equalTo(new RtGithub(new FakeRequest()))
         );
     }
+
+    /**
+     * RtGithub can connect easily to another Github [enterprise] server
+     */
+    @Test
+    public void connectsToAnotherGithubServer() throws Exception {
+        final RtGithub github = new RtGithub(URI.create("http://localhost:12345"));
```
Comment (pinaf) [inline src/test/java/com/jcabi/github/RtGithubTest.java]: @perfa we can inline this variable.

File: src/main/java/com/jcabi/github/RtGithub.java [hunk 2]
```diff
@@ -87,14 +88,17 @@
         Manifests.read("JCabi-Date")
     );
 
+    public static final String DEFAULT_GITHUB_HOST = "https://api.github.com";
```
Comment (pinaf) [inline src/main/java/com/jcabi/github/RtGithub.java]: @perfa needs javadoc

File: src/main/java/com/jcabi/github/RtGithub.java [hunk 3]
```diff
@@ -106,7 +110,7 @@
      * @since 0.4
      */
     public RtGithub() {
-        this(RtGithub.REQUEST);
+        this(getRequest(DEFAULT_GITHUB_HOST));
```
Comment (pinaf) [inline src/main/java/com/jcabi/github/RtGithub.java]: @perfa `RtGithub.getRequest`

File: src/main/java/com/jcabi/github/RtGithub.java [hunk 4]
```diff
@@ -118,8 +122,39 @@ public RtGithub() {
...
+        this(URI.create(DEFAULT_GITHUB_HOST), token);
+    }
+
+    /**
+     * Public ctor, for anonymous access to Github Enterprise.
+     * @param uri URI of Github server
+     */
+    public RtGithub(
+        @NotNull(message = "uri can't be NULL") final URI uri) {
+        this(getRequest(uri.toString()));
```
Comment (pinaf) [inline src/main/java/com/jcabi/github/RtGithub.java]: @perfa `RtGithub.getRequest`

File: src/main/java/com/jcabi/github/RtGithub.java [hunk 5]
```diff
@@ -134,12 +169,14 @@ public RtGithub(
...
      * Public ctor, for authentication with OAuth2 token.
+     * @param uri URI of Github server
      * @param token OAuth token
      */
     public RtGithub(
+        @NotNull(message = "uri can't be NULL") final URI uri,
         @NotNull(message = "token can't be NULL") final String token) {
         this(
-            RtGithub.REQUEST.header(
+            getRequest(uri.toString()).header(
```
Comment (pinaf) [inline src/main/java/com/jcabi/github/RtGithub.java]: @perfa `RtGithub.getRequest`

File: src/main/java/com/jcabi/github/RtGithub.java [hunk 6]
```diff
@@ -118,8 +122,39 @@ public RtGithub() {
     public RtGithub(
         @NotNull(message = "user name can't be NULL") final String user,
         @NotNull(message = "password can't be NULL") final String pwd) {
+        this(URI.create(DEFAULT_GITHUB_HOST), user, pwd);
```
Comment (pinaf) [inline src/main/java/com/jcabi/github/RtGithub.java]: @perfa `RtGithub.DEFAULT_GITHUB_HOST`

File: src/main/java/com/jcabi/github/RtGithub.java [hunk 7]
```diff
@@ -118,8 +122,39 @@ public RtGithub() {
...
+    }
+
+    /**
+     * Public ctor, for authentication with OAuth2 token.
+     * @param token OAuth token
+     */
+    public RtGithub(
+        @NotNull(message = "token can't be NULL") final String token) {
+        this(URI.create(DEFAULT_GITHUB_HOST), token);
```
Comment (pinaf) [inline src/main/java/com/jcabi/github/RtGithub.java]: @perfa `RtGithub.DEFAULT_GITHUB_HOST`

**Comment by pinaf** [comment]
@perfa a few comments

**Comment by pinaf** [comment]
@perfa 
>
[ERROR] /src/main/java/com/jcabi/github/RtGithub.java[263]: Line has trailing spaces. (RegexpSinglelineCheck)
[ERROR] /src/main/java/com/jcabi/github/RtGithub.java[263]: Should contain one indentation space (MultilineJavadocTagsCheck)

**Comment by perfa** [comment]
@pinaf fixed. Took me a while, I had maven issues preventing me from running qulice. Finally got it working. Should I squash the fixup commits, or are they fine like they are?

**Comment by pinaf** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by perfa** [comment]
@yegor256 @pinaf That Rultor failure looks a bit like a flakey test, it's unrelated to my change (the URL it was using looks well-formed), and the only change on master looks like documentation updates.

**Comment by yegor256** [comment]
@rultor try to merge again

**Comment by perfa** [comment]
Strange... It gets  `"name already exists on this account"` when it tries to create a repo. Seems like the randomly generated repo name shouldn't really manage to overlap with some previous test-run, but the behavior it's testing is unchanged, apart from how the URL is created which is working. Is this something you've had problems with before @yegor256 ?

**Comment by yegor256** [comment]
@perfa nope, I haven't seen this before. keep this in mind: http://www.yegor256.com/2015/01/15/how-to-cut-corners.html

**Comment by pinaf** [comment]
@perfa ping

**Comment by pinaf** [comment]
@perfa if you gave up, please close the PR. Thanks!

**Comment by perfa** [comment]
@pinaf Yeah, we've chosen another route. Will close.

**Comment by dmarkov** [comment]
@pinaf Thank you, I have added **27 mins** to your account in payment/transaction "`AP-1SB47903T3909415N`", time consumed: 1780 hours and 18 mins

review comments (c=12) added as a bonus

+27 to your rating, your total score is [+6467](http://www.netbout.com/b/35128?open=rating)

**Pull request closed without merging**